drivers/flash/flash_stm32_ospi.c

```diff
@@ -976,8 +981,7 @@ static int flash_stm32_ospi_write(const struct device *dev, off_t addr,
 		}
 	}
 
-	cmd_pp.Address = addr;
-	cmd_pp.AddressSize = stm32_ospi_hal_address_size(dev);
```

Comment:
Not every flash IC works with a 32-bit address size. Not sure why the setting of the size is removed here...

Reply:
In case of OPI/STR transfer, the cmd.AddressSize must be HAL_OSPI_ADDRESS_32_BITS 
and stm32_ospi_hal_address_size gives HAL_OSPI_ADDRESS_24_BITS  which fails when reading:
```
Data read does not match data written!!                                         
000ff000 wrote 55 read ff MISMATCH                                              
000ff001 wrote aa read ff MISMATCH                                              
000ff002 wrote 66 read ff MISMATCH                                              
000ff003 wrote 99 read ff MISMATCH 
```

Reply:
I don't know what Flash you use, but MX25LM51245G does support 24 and 32 bits address sizes depending on the configuration. Winbond's W25Q128JV supports only 24 bits address size.

Therefore, it is not so much Mode dependent rather than Flash IC-specific.

Considering some of your previous comments your flash works with 4-byte opcodes and that requires 4-byte address width (_otherwise it is a configuration error_).

Currently, the address width is taken from SFDP,BFP:

> `spi_nor_process_bfp_addrbytes(dev, jesd216_bfp_addrbytes(bfp));`

Please check your SFDP table and enable `four-byte-opcodes` in the DTS.

The alternative is to allow to set address width in the DTS and prioritize it over SFDP, BFP.

Reply:
@FRASTM, ping.

Alternative: 

Quad and dual flashes might not work with a 32-bits address size.
ospi_prepare_cmd enforces 32-bits address size. Please use 24-bits for non OSPI_OPI_MODE.

Reply:
ok